OCM-9613 | feat: Allow billing account update via the cluster edit command

cristianoveiga opened the pull request.

ciaranRoche reviewed (commented):
This LGTM, just to confirm is there any blocker keeping this in draft?

cristianoveiga commented:
> This LGTM, just to confirm is there any blocker keeping this in draft?

Thanks @ciaranRoche . We have been asked to review this change with the UX team.

I will resume the work on this whenever we get their input.

ciaranRoche commented:
@cristianoveiga one worry is this will get out of date, and result in more work on your end to keep up to date. Do we have an idea of how long it will take for UX review

abhgupta commented on cmd/edit/cluster/cmd.go:
```diff
@@ -160,6 +163,13 @@ func init() {
+		&args.billingAccount,
+		"billing-account",
+		"",
+		"Account used for billing subscriptions purchased via the AWS marketplace",
```
If this is about ROSA, customers don't directly purchase ROSA through the AWS marketplace. They do so through the AWS console as ROSA is offered as a 1st party service through AWS

lnagendr replied:
This could be a better wording "Account used for billing subscriptions purchased via the AWS console for ROSA"

abhgupta commented on cmd/edit/cluster/cmd.go:
```diff
@@ -651,6 +661,76 @@ func run(cmd *cobra.Command, _ []string) {
+			r.Reporter.Infof(
+				"The selected AWS billing account is a different account than your AWS infrastructure account." +
+					"The AWS billing account will be charged for subscription usage. " +
+					"The AWS infrastructure account will be used for managing the cluster.",
```
No space between the first and second sentence.
This needs rewording - "The AWS infrastructure account will be used for managing the cluster."

lnagendr replied:
Can be reworded as "The AWS infrastructure account contains the ROSA infrastructure".

abhgupta commented:
Additional review comments that I could not figure out where to comment in the file changes.
1. The "Billing Account" prompt should be "Update billing account"
2. Instead of "default", we should highlight the AWS account as the "current" billing account that is shown within parenthesis.
3. The text message once the user switches their AWS billing account should be entirely rewritten.

cristianoveiga commented:
> Additional review comments that I could not figure out where to comment in the file changes.
> 
> 1. The "Billing Account" prompt should be "Update billing account"
> 2. Instead of "default", we should highlight the AWS account as the "current" billing account that is shown within parenthesis.
> 3. The text message once the user switches their AWS billing account should be entirely rewritten.

I've addressed items 1 and 2 above.

jseseCCS commented on cmd/create/cluster/cmd.go:
```diff
@@ -3348,14 +3347,6 @@ func clusterConfigFor(
 
-func validateBillingAccount(billingAccount string) error {
-	if billingAccount == "" || !ocm.IsValidAWSAccount(billingAccount) {
-		return fmt.Errorf("billing account is invalid. Run the command again with a valid billing account. %s",
```
(assuming it's the acct _number_ that's not valid,) consider changing to: Billing account number is not valid. Rerun the command with a valid billing account number.

Also, good error messages provide user help if the retry doesn't work. what can the user do if they still get an error?

cristianoveiga replied:
Hi @jseseCCS , I've updated the error message with the one you suggested.

We are planning to add a `rosa list billing-accounts` command (tracked via OCM-9642).

I was thinking we could revisit this error message after we implemented that.

The final result would be something like:

```
Billing account number is not valid. Rerun the command with a valid billing account number. 
To see the list of billing account options, run the rosa list billing-accounts command.
```

Thoughts?

lnagendr replied:
Sounds good!

lnagendr commented:
This LGTM. I don't have any additional comments.

cristianoveiga commented:
/retest

cristianoveiga commented:
/retest

cristianoveiga commented:
> > This LGTM, just to confirm is there any blocker keeping this in draft?
> 
> Thanks @ciaranRoche . We have been asked to review this change with the UX team.
> 
> I will resume the work on this whenever we get their input.

Hi @ciaranRoche , I've addressed the PR comments and we have received green light from @lnagendr .

Can you please take another look at this and merge it if you don't have any additional comments?

jseseCCS commented on cmd/create/cluster/cmd.go:
```diff
@@ -721,7 +720,7 @@ func initFlags(cmd *cobra.Command) {
 		"billing-account",
 		"",
-		"Account used for billing subscriptions purchased via the AWS marketplace",
+		"Account used for billing subscriptions purchased via the AWS console for ROSA",
```
hello!

sorry to crash the PR but, for some reason, I got email notifs about this content, so here I am. I'm Jocelyn, writer for OCM.

if I were to edit this content, i'd tell you that, for translation reasons, our style guides prohibit us from using "via." "through" is better.

cristianoveiga replied:
Hi @jseseCCS , no apologies needed, thanks for your input.

I will review and address your comments.

lnagendr replied:
+1

cristianoveiga replied:
updated: `via` to `through`

jseseCCS commented on cmd/create/cluster/cmd.go:
```diff
@@ -1150,20 +1149,20 @@ func run(cmd *cobra.Command, _ []string) {
 				if billingAccount != awsCreator.AccountID {
 					r.Reporter.Infof(
-						"The selected AWS billing account is a different account than your AWS infrastructure account." +
+						"The selected AWS billing account is a different account than your AWS infrastructure account. " +
```
suggested rewrite: 

The AWS billing account you selected is different from your AWS infrastructure account.

OR

Select your AWS infrastructure billing account. The AWS account you selected is not your AWS infrastructure account. 

OR

Select your AWS infrastructure billing account. The AWS account you selected is not the correct account.

lnagendr replied:
+1 "The AWS billing account you selected is different from your AWS infrastructure account."

cristianoveiga replied:
updated as suggested

jseseCCS commented on cmd/create/cluster/cmd.go:
```diff
@@ -1150,20 +1149,20 @@ func run(cmd *cobra.Command, _ []string) {
+						"The selected AWS billing account is a different account than your AWS infrastructure account. " +
 							"The AWS billing account will be charged for subscription usage. " +
-							"The AWS infrastructure account will be used for managing the cluster.",
+							"The AWS infrastructure account contains the ROSA infrastructure.",
```
the _account_ contains the infrastructure? or does the thing that the account is _for_ contain the infrastructure?

lnagendr replied:
@jseseCCS it's the latter.

jseseCCS commented on cmd/edit/cluster/cmd.go:
```diff
@@ -160,6 +163,13 @@ func init() {
+		&args.billingAccount,
+		"billing-account",
+		"",
+		"Account used for billing subscriptions purchased via the AWS console for ROSA",
```
via

lnagendr replied:
+1

jseseCCS commented:
hey, guys!

a general note about [error messages](https://www.ibm.com/docs/en/ibm-style?topic=format-messages#error-messages): Whenever possible, tell users **why** the problem occurred and **how** to fix or diagnose the problem.

jseseCCS commented:
let me know if i can be of any more help. I'm just a ping/@ away.

cristianoveiga commented:
> let me know if i can be of any more help. I'm just a ping/@ away.

Thanks!

cristianoveiga commented:
/retest

hunterkepley commented on cmd/edit/cluster/cmd.go:
```diff
@@ -695,6 +705,77 @@ func run(cmd *cobra.Command, _ []string) {
+			os.Exit(1)
+		}
+		if billingAccount != "" && !ocm.IsValidAWSAccount(billingAccount) {
+			r.Reporter.Errorf("Billing account number is not valid. Rerun the command with a valid billing account number")
```
NP: Might be good to include the billing account number back to the user such as `Billing account number '%s' is not valid...`

hunterkepley replied:
* If OCM does not bubble this up anyway

hunterkepley replied:
Also make `Provided billing account number...` would be better, but again not needed

cristianoveiga replied:
Thanks, @hunterkepley I have updated the PR with your suggestions.

hunterkepley reviewed (approved):
LGTM, nitpick on UX with error displayed to user, but maybe OCM bubbles up what I'm asking

hunterkepley commented:
Validation + Tests look good to me

cristianoveiga commented:
/retest

gdbranco commented:
/lgm
/approve

gdbranco commented:
/lgtm
